File: peer/src/context.rs
use log::info;
use std::{
    error::Error,
    sync::{Arc, Mutex},
    thread::JoinHandle,
};

use common::{types::*, util::system_time_msec};
use dittolive_ditto::prelude::*;

use crate::HeartbeatCtx;
use crate::consumer::PeerConsumer;

pub struct PeerContext {
    pub id: PeerId,
    pub ditto: Ditto,
    pub coord_addr: Option<String>,
    pub coord_doc_id: Option<DocumentId>,
    pub coord_info: Option<CoordinatorInfo>,
    // Keep a copy of our last transport config so we can modify and re-set it.
    pub transport_config: Option<TransportConfig>,
    #[allow(dead_code)]
    pub hb_doc_id: Option<DocumentId>,
    pub hb_ctx: Option<HeartbeatCtx>,
    pub hb_thread: Option<JoinHandle<Result<(), std::io::Error>>>,
    #[allow(dead_code)]
    pub start_time_msec: u64,
    pub local_ip: String,
    pub state: Arc<Mutex<PeerState>>,
    pub peer_collection: Option<Arc<Mutex<Collection>>>,
    pub peer_consumer: Option<PeerConsumer>,
}

impl PeerContext {
    pub fn new(device_name: &str, ditto: Ditto, local_ip: &str) -> Self {
        Self {
            id: random_peer_id(Some(&device_name)),
            ditto,
            coord_addr: None,
            coord_doc_id: None,
            coord_info: None,
            transport_config: None,
            hb_doc_id: None,
            hb_ctx: None,
            hb_thread: None,
            start_time_msec: system_time_msec(),
            local_ip: local_ip.to_string(),
            state: Arc::new(Mutex::new(PeerState::Init)),
            peer_collection: None,
            peer_consumer: None,
        }
    }

    pub fn get_plan(&self) -> Option<ExecutionPlan> {
        let ci = self.coord_info.as_ref()?;
        ci.execution_plan.clone()
    }

    pub fn state_transition(
        &mut self,
        existing: Option<PeerState>,
        new: PeerState,
    ) -> Result<(), Box<dyn Error>> {
        let mut state = self.state.lock().unwrap();
        assert!(existing.is_none() || existing.unwrap() == *state);
        info!("--> state_transition: {:?} -> {:?}", state, new);
        *state = new;
        Ok(())
    }
}

File: container-mesh/ts/src/context.ts
import { Ditto, DocumentID, TransportConfig } from "@dittolive/ditto"
import { CoordinatorInfo, PeerId, PeerState, SerializedPeer } from "./types"
import { COORD_COLLECTION_NAME } from "./default"

export class PeerContext {
    id: PeerId
    ditto: Ditto | null
    coord_addr: string
    coord_port: number
    coord_doc_id: DocumentID | null
    coord_collection: string
    coord_info: CoordinatorInfo | null
    transport_config: TransportConfig | null
    hb_doc_id: DocumentID | null
    hb_timer: NodeJS.Timeout | null
    //start_time_msec: number,
    local_addr: string
    local_port: number
    state: PeerState

    constructor(id: PeerId, ditto: Ditto, coord_addr: string, coord_port: number,
                bind_addr: string, bind_port: number) {
        this.id = id
        this.ditto = ditto
        this.coord_addr = coord_addr
        this.coord_port = coord_port
        this.coord_doc_id = null
        this.coord_collection = COORD_COLLECTION_NAME
        this.coord_info = null
        this.transport_config = null
        this.hb_doc_id = null
        this.hb_timer = null
        this.local_addr = bind_addr
        this.local_port = bind_port
        this.state = PeerState.Init
    }

    stateTransition(from: PeerState, to: PeerState): void {
        if (this.state != from) {
            throw new Error(`Invalid state transition from ${PeerState[from]} to ${PeerState[to]}`)
        }
        console.debug(`--> stateTransition ${PeerState[from]} -> ${PeerState[to]}`)
        this.state = to
    }

    toSerializedPeer(): SerializedPeer {
        return {
            peer_id: this.id,
            peer_ip_addr: this.local_addr,
            peer_port: this.local_port,
            state: PeerState[this.state]
        }
    }
}
